### DISCUSSION TITLE
inout methods on a class are created like foo() & -> void. What is that ampersand doing there?

### DISCUSSION THREAD

**Discussion by tomFlowee** [hsutter/cppfront#901] [Suggestions]
As seen in various bugreports, users are having problems getting their head around this one. I would argue it is because there are more usecases that need to be handled differently.

Bugreport #900, #839 and #792 are the ones I found.

Todays state;

cppfont has some cool code to determine last usage of a variable. For things like unique_ptr to implicitly work this has been used to add a std::move and do the thing that people expect.

This creates a problem, however, for instances of self-made classes.

Imagine:
```
A: type = {
    connect: (inout this, s: std::string) = {
    }
}

main: (args) -> int = {
    a: A = ();
    a.connect("bar");
}
```

The variable `a` is allocated on-stack. The core language has a very clear set of rules on the lifetime. Which raises the question, when is it ever Ok to generate a 'std::move' wrapper for usage of this variable?

If 'connect' would be static, the above code would compile and actually the lifetime of `a` is still going to be till the end of the scope of main (as expected).

*So why would it be useful to add a std::move to the call of connect() ?*

People have stated I can work around this by explicitly discarding the instance of 'a' at the end. `_ = a;`. It does feel like that misses the point, I have no intention of discarding 'a', it is supposed to live till the end of main(). ([usecase](https://github.com/hsutter/cppfront/issues/900#issuecomment-1865033787)

Maybe someone can link to the underlying reason for the existance of the std::move that is currently generated, so we can find a way to make all usecases work.

Thanks for reading!

**Comment by gregmarr** [comment]
This isn't really connected to UFCS.  You get the same behavior whether you call with member syntax or non-member syntax.

```
A: type = {}

connect: (inout arg: A, s: std::string) = {}

main: (args) -> int = {
    a: A = ();
    connect(a, "bar");
}
```

```
    A a {}; 
    connect(std::move(a), "bar");
```

**Comment by hsutter** [comment]
Thanks! This is a good question.

As background, please read ["Design note: Explicit discard"](https://github.com/hsutter/cppfront/wiki/Design-note%3A-Explicit-discard)

Consider your example, adding a comment:

>```
> A: type = {
>     connect: (inout this, s: std::string) = {   // <=== note the INOUT here
>     }
> }
> 
> main: (args) -> int = {
>     a: A = ();
>     a.connect("bar");
> }
>```

If that `inout` were omitted (`in` is the default), this program compiles without error.

This highlights the reason the above code is getting the error: It is _modifying `a`_, then never looking at `a` again... the "out" value of `a` is dropped on the floor. It's the same case as if a returned value was never looked at and dropped on the floor, but in this case the returned value is the "out" value of `this` object.

My thinking is that making this break is a feature, not a bug. Looking at the code above, is it correct that the code modified `a`, but then immediately destroys it without using the new value it just cause `a` to have?

**Reply by tomFlowee** [reply]
>  It is modifying a, then never looking at a again.

Well, kinda.

I mean, you are making a lot of assumptions in that statement.
If 'a' was a networkmanager class and calling 'bind' on it while passing it a function pointer then your assumptions would be wrong.

The more accurate statement is that THIS METHOD doesn't call methods anymore. That is not at all like the instance is "dropped on the floor". I can come up with a million useful usecases here.

While in the example it is true that the next thing that happens is that the instance destructor will be called, in real code that may not be the case. Some app-wide eventloop may be started causing the lifetime of this object to be 'forever'.

Next to that, this feels like the crappy IDE experience where it loudly proclaims I'm wrong about something, because I'm only halfway down typing this method and if I continue then the warnings from the IDE go away.
The comparison stops when you realize that the IDE only warns in such cases. You can compile and run your unfinished code just fine. But not in cppfront because you make assumptions that do not hold for a lot of cases.


Edit: maybe I misunderstand the 'inout' concept.
I intend to created a "standard" cpp method. i.e. `void foo(const std::string &bar);`
When I just write `foo: (bar: std::string)` it is static. So I needed to add 'this' there to make it normal. Maybe there is a PEBCAK issue I'm not following...

**Reply by hsutter** [reply]
> If 'a' was a networkmanager class and calling 'bind' on it while passing it a function pointer then your assumptions would be wrong.

Can you give a short example?

> The more accurate statement is that THIS METHOD doesn't call methods anymore.

But because this only happens with local-scope objects, "this function" is all there is. We can see lexically that nobody mentions the object's name anymore (to call functions or do anything else) between that last call and when the object is destroyed at end of scope.

> While in the example it is true that the next thing that happens is that the instance destructor will be called, in real code that may not be the case. Some app-wide eventloop may be started causing the lifetime of this object to be 'forever'.

I'm not sure how could that happen for a local object, can you give a short example?

Again, in case it's helpful: Note that injecting `std::move` on the last use does not necessarily actually move from the object (it only moves if the function being called actually has a move overload), and never destroys it (the destructor call is unchanged, it doesn't happen until the end of its scope as always).

> I intend to created a "standard" cpp method. i.e. `void foo(const std::string &bar);`

Clarifying question: Note that in today's code as you wrote it here, the default is a _non-const_ member function. Is that what you wanted?

In Cpp2, the default would be a _const_ member function, so you would write it like this which adds the explicit `this` and passes `bar` as an "in" parameter which is equivalent to the reference to const: `foo: (this, bar: std::string)`.

If you really wanted it to be a _non-const_ member function, though, just change `this` to `inout this`.

**Reply by gregmarr** [reply]
> If you really wanted it to be a non-const member function, though, just change this to inout this.

When you do that, you get a `&` this modifier on it, and then you can't call it as a definite-last-use in Cpp2 as it gets `std::move()` applied to the object which converts it to an rvalue, and thus the member function can't be called.

> But because this only happens with local-scope objects, "this function" is all there is. We can see lexically that nobody mentions the object's name anymore (to call functions or do anything else) between that last call and when the object is destroyed at end of scope.

https://cpp2.godbolt.org/z/7f14s7ea7

```
my_raii_class : type = {
    operator=: (out this, _: int) = {}
    operator=: (move this) = {
      // tear down the database connection
    }

    init: (inout this, var : std::string) = {
      // do something external such as setting up a database connection using a singleton
      myvar = var;
    }

    myvar : std::string = ();
}

doStuff : () = {
    raii : my_raii_class = 1;

    if (foo()) {
        raii.init("");
    } else {
        raii.init("");
    }

    // do stuff that depends on raii having been constructed but not yet destructed()
}
```

Because the two `init()` calls are both definite-last-uses, they get `std::move()` applied, and you can't call `init()` on an `raii&&` because it's lowered as 

```
public: auto init(cpp2::in<std::string> var) & -> void;
```

**Reply by tomFlowee** [reply]
> Can you give a short example?

since I just finished my test app, I can ([git](https://codeberg.org/zander/cpp2-versionCheck/src/branch/master/main.cpp2#L9-L10));

```
main: (args) -> int = {
    app: QCoreApplication = (args.argc, args.argv);
    digger : Digger = Digger();
    digger.connect("localhost", 1235);
    return app.exec();
}
```

> But because this only happens with local-scope objects, "this function" is all there is. We can see lexically that nobody mentions the object's name anymore (to call functions or do anything else) between that last call and when the object is destroyed at end of scope.

As you can see above in the example, between the 'connect' and the end of the scope is all the time in the world.

> Note that injecting std::move on the last use does not necessarily actually move from the object

I think we figured out that the move is indeed something that can stay unchanged, the problem is the addition of that ampersand in the generated function delcaration (of connect) in cppfront. THAT causes the issues. Remove that and the problems go away.

> Note that in today's code as you wrote it here, the default is a non-const member function. Is that what you wanted?

That is what I wanted. A non-const method **that doesn't have that weird ampersand**. And as you can see from the 5 different tries, I don't think it is that easy to get it. AFAICT cppfront refuses to create the code I'm aiming to get.

**Reply by JohelEGP** [reply]
> AFAICT cppfront refuses to create the code I'm aiming to get.

By design, if the code you're aiming to get has to _not_ have `_ = digger` after the return statement.
As I showed at the beginning: <https://github.com/hsutter/cppfront/issues/900#issuecomment-1865014236>.

That is a weird thing to say,
like "I want to drive my Tokyo, but my car refuses to operate my way",
and your way includes not turning on the car with your keys.

**Reply by gregmarr** [reply]
Well, we are still designing the car here.  (By the way, I don't turn on my car with my keys.  They just need to be in the car.)

The current design is "if you call a function that mutates an object, you MUST after that somehow use the object again in the same function after  you call it". This ignores things like "this object has a constructor that registers it with an external system that observes the changes to that object, and a destructor that deregisters the object".

I had a thought that maybe you can get around this issue by using a smart pointer.

```
    raii := new<my_raii_class>(1);

    if (foo()) {
        raii.get()*.init("");
    } else {
        raii.get()*.init("");
    }
```

This actually works.  If you have a smart pointer to an object, you can call mutating functions on it as the last thing in the function before the object` is destroyed without having to tell the compiler. 

It's not pretty, however.  I'm also a bit unsure about having to transform  `->` to `.get()*.` when moving from cpp1 to cpp2.  I thought we had a discussion about this, but I can't find it right now.

FYI, I just found another discussion on this topic.  #231

**Reply by JohelEGP** [reply]
#231 has been resolved, along with other related issues listed by #466.

**Reply by gregmarr** [reply]
That may be the case (even though it's still open), but there is still a lot of useful discussion in that issue.

**Reply by tomFlowee** [reply]
```
    raii := new<my_raii_class>(1);

    if (foo()) {
        raii.get()*.init("");
    } else {
        raii.get()*.init("");
    }
```

That's an interesting usage, it downcasts stuff to raw pointers to make it work.
Which is not very cpp2 style, but the cpp2 style isn't very easy to for me to get working. Some of the problems from this discussion may go away if I could get the setup working where I can use classes that are not on stack.

Anyone knows how to rewrite the hopefully obvious intention here to actually make stuff compile?

```
Item : type = {
    setNext: (inout this, next: std::unique_ptr<Item>) = {
        m_next = next;
    }
    private m_next: std::unique_ptr<Item> = ();
}

main: (args) -> int = {
     firstItem:= new<Item>();
     secondItem:= new<Item>();
     firstItem.setNext(secondItem);
}
```

Not sure how to store that private instance, any type is fine with me. But how does cpp2 prefer this?

**Reply by JohelEGP** [reply]
You need 2 things (<https://cpp2.godbolt.org/z/Wxqdanjae>):
1. Make `next` a `move` parameter.
2. Dereference `firstItem` before calling the member function on the stored object.

**Reply by tomFlowee** [reply]
Thanks!

The "leaking through" of hidden (mandatory) things like stuff suddenly becoming a smart pointer, the mandatory move semantics, the topic of this specific discussion and as a random point:  "you can't use `->`, please use `*.` are things that *need work*.

I understand the goals, but the learning cost is too great and long explanations non-withstanding, they really only make sense if you are actually the author of this conversion layer. For the rest of us they are just new bullets to aim at our feet.

Fast forward 5 years, I expect MS to not have backed down on the government threats, so these problems won't go away in cppfront. But we'll likely see a spin-off that takes all the nice things without the breaking "cpp1 is bad!" stuff and that may be actually useful for the larger audience.

Cheers!

**Reply by iprtel** [reply]
Please consider this:
```
ResultSet: type = {
    counter : int = 0;
    next : (inout this) -> bool = {
        counter++;
        return counter == 0;   // Always returns false
    }
    value : ( this) -> int = {
        return counter;
    }
}

test : () = {
    a:_ = new<ResultSet>();
    _ = a*.next();

    b:ResultSet = ();
    _ = b.next();
    _ = b;
}

test_if : () = {
    a:_ = new<ResultSet>();
    if (a*.next()) {
    }

    b:ResultSet = ();
    if (b.next()) {
        _ = b.value(); // [1]  
    }
    //_ = b; [2]
}

test_while: () = {
    a:_ = new<ResultSet>();
    while (a*.next()) {
    }

    b:ResultSet = ();
    while (b.next()) {
    }
}

main : () = {
    test();
    test_if();
    test_while();
}```

`ResultSet` represents the result of a query. Initially it points 'before' the actual result, with the first call to `next` it is internally moved to the first result if there is one. `next()` will return true only if a valid result is available.

This can be used in different ways:
1. if the caller is only interested if there is a valid result one call to `b.next()` will be sufficient.
2.  if the caller expects at most one result `if (b.next())` will do the trick. (Why is it sufficient to add the call in one branch that is never called?)
3. if the caller requires all results loop over the ResultSet

How can I explain the different handling of 'b' in all these cases where the pointer version can be used the same all the time?
Can I have a 'next(mutable this)` where the mutable indicates that the object might change, even if the caller is not interested in the change?

**Reply by JohelEGP** [reply]
> (Why is it sufficient to add the call in one branch that is never called?)

The analysis doesn't consider execution.
What matters is that the use in the condition isn't the definite last use.
For that, a use after the condition (within the branch, or after the selection statement) suffices.

> How can I explain the different handling of 'b' in all these cases where the pointer version can be used the same all the time?

The unique pointer is still moved.
But it is dumb, and always literally gives you what it points to.
It doesn't even [propagate `const`](https://en.cppreference.com/w/cpp/experimental/propagate_const).

If you switch to `std::optional`, which does forward rvalueness,
you then get an error just like the unwrapped case: <https://cpp2.godbolt.org/z/qd3n76bbb>.

> Can I have a 'next(mutable this)` where the mutable indicates that the object might change, even if the caller is not interested in the change?

It was decided that there wouldn't be a keyword for this in a function declaration (somewhere in #231, I think).

**Reply by iprtel** [reply]
Thanks for your responses at this time of the year.

> The analysis doesn't consider execution.
What matters is that the use in the condition isn't the definite last use.
For that, a use after the condition (within the branch, or after the selection statement) suffices.

I'm afraid lots of people will add `CPP2_DISCARD(x);` for every variable that the compiler complains at the end of a function.

> The unique pointer is still moved.
But it is dumb, and always literally gives you what it points to.
It doesn't even [propagate const](https://en.cppreference.com/w/cpp/experimental/propagate_const).
> If you switch to std::optional, which does forward rvalueness,
you then get an error just like the unwrapped case: https://cpp2.godbolt.org/z/qd3n76bbb.

That is interesting about the unique pointer. Could `new` return a std::optional of a reference type that cannot be set to null? 
Then I could do a null check once after the new and then never bother again. That could also fix the issue here.

**Reply by gregmarr** [reply]
> That's an interesting usage, it downcasts stuff to raw pointers to make it work.
> Which is not very cpp2 style

Yeah, I'm not promoting that as a great method, just evaluating what's possible with the current analysis to see if the current analysis is appropriate.

**Comment by JohelEGP** [comment]
> People have stated I can work around this by explicitly discarding the instance of 'a' at the end. `_ = a;`. It does feel like that misses the point, I have no intention of discarding 'a', it is supposed to live till the end of main().

Discarding isn't the same as destroying, so it doesn't affect the lifetime of `a`.
You're just explicitly stating that you're not using an "out" aspect of it, as Herb says.
In the case of calling an `inout` member, by discarding afterwards,
the move on last use becomes part of the discard, and that `std::move`, by itself, does nothing.

**Comment by hsutter** [comment]
Right, "discard" refers to _the value_.

This terminology (and syntax) does seem to follow the prior art in other languages mentioned in the design note.

Note that I didn't have C#'s experience in mind -- this is not C# envy or anything. 😉  I independently arrived at the `_ = f();` discard syntax because Cpp2 already allowed `_` as "don't care" in lots of places, and I wanted a way to say "I don't care about the result of `f()`," so `_ = f();` seemed a natural way to write it, connoting 'assign the result to nothing/I-don't-care.' I think this is an example of convergent design/evolution, because I subsequently found C# had made the same choice, and presumably it was for similar reasons. Then when considering `inout` parameters, it seemed natural to use the same `_ = obj;` to say "see, I'm explicitly not looking at `obj`'s value."

**Comment by hsutter** [comment]
BTW, I've just got to say that, even though ChatGPT does sometimes make mistakes(*), it's really impressive... "magical" as defined by Clarke, given that AFAIK no one, not even the AI designers, really deeply groks how it generates the output it does.

Considering that Cpp2 is a small personal project that only existed on the web in the past year, and yet I get this (using ChatGPT 4):

<p align="center"><img width="500" src="https://github.com/hsutter/cppfront/assets/1801526/c81f7014-7693-4e9c-b58e-6adc7dd55d7f"></p>

The link is the above-linked design note... cool.

Update: Here's another...

<p align="center"><img width="500" src="https://github.com/hsutter/cppfront/assets/1801526/b538e247-8e3f-4d9f-bc10-245357442f48"></p>

And this is just general-purpose garden-variety ChatGPT... it's not a dedicated Copilot that could have domain-specific training, and as far as I know it's not using any special plugin/gpt.

---

(*) The first time I asked it what cppfront was, it gave a completely correct synopsis except that it said the compiler was written by Bjarne Stroustrup. 😆 In writing this example I tried it again, and this time it got everything correct.

**Comment by tomFlowee** [comment]
I have the feeling we're actually mostly in agreement, but this is about a very different thing...

To put it in the most simple way; Why does this code compile, but my example does not?

```
#include <fstream>
main: (args) -> int = {
    a: std::ofstream = ();
    a.open("filename");
    // ^ a non-const method call that likewise returns
    //  nothing. Exactly like my 'connect' method intended to do.
}
```

I want my example to behave the same way, I call a non-const method on an object that has plenty of logic on it and I want the destructor to only be called at the end of the scope.

Ok, goal:

 `public: auto connect(cpp2::in<std::string> s) -> void;`

try 1: `A: type = { connect: (this, s: std::string) = { } }`
gives: `public: auto connect(cpp2::in<std::string> s) const& -> void;`

try 2:  `A: type = { connect: (in this, s: std::string) = { } }`
gives the same.

try 3: `A: type = { connect: (out this, s: std::string) = { } }`
gives:  `public: auto connect(cpp2::in<std::string> s) -> A& ;`
that looks odd...

try 4: `A: type = { connect: (inout this, s: std::string) = { } }`
gives:  `public: auto connect(cpp2::in<std::string> s) & -> void;`

try 5: `A: type = { connect: (s: std::string) = { } }`
gives: `public: static auto connect(cpp2::in<std::string> s) -> void;`

I'm out of creativity; how do I get a normal plain and simple void-returning, non-const method that is a member on a class? I think that is the same question as OP. How do I get a normal simple method I can call without qualifying as "dropping the result on the floor".

**Reply by JohelEGP** [reply]
The same happens with
```Cpp2
    a: std::string = ();
    a.clear();
```

It looks like it's the added ref qualifier `&` on Cpp2 `inout` member functions.
If I use a Cpp1 explicit object parameter, it also doesn't work (<https://cpp2.godbolt.org/z/nWrzGMar1>):
```Cpp2
struct my_string {
  void clear(this my_string&) { }
};
main: () -> int = {
    a: my_string = ();
    a.clear();
}
```

The trail of commits for emitting `&`s is 5a252cc9e3210451ad590c47458d470b9bcdbe60 and 209f6d56b4819eeade9c85fae59d739b8e55eb10.

**Reply by JohelEGP** [reply]
From <https://en.cppreference.com/w/cpp/language/overload_resolution#Additional_rules_for_member_function_candidates>:
> 2) rvalues can be bound to non-const implicit object parameter (unless this is for a ref-qualified member function)(since C++11) and do not affect the ranking of the implicit conversions.

**Reply by gregmarr** [reply]
Yes, you get the same thing here:

```
auto connect(A& arg, std::string s) -> void{}

auto main(int const argc_, char** argv_) -> int{
    A a {}; 
    connect(std::move(a), "bar");
}
```

This fails to compile with the `std::move()`.

**Reply by hsutter** [reply]
> How do I get a normal simple method I can call without qualifying as "dropping the result on the floor".

Make it `in this` (the default, which can be spelled just `this`).

I think what you're saying is that you want to write a function that _modifies_ `this` object, but whose side effect on the value of `this` object can be silently ignored?

---

> ```
> #include <fstream>
> main: (args) -> int = {
>     a: std::ofstream = ();
>     a.open("filename");
>     // ^ a non-const method call that likewise returns
>     //  nothing. Exactly like my 'connect' method intended to do.
> }
> ```

> To put it in the most simple way; Why does this code compile, but my example does not?

It's because Cpp2-authored types are trying to be stricter about non-const member functions being `&`-qualified. @JohelEGP gave more details above.

That's a good example, though, because it shows that the result can be surprising.

FWIW, I think I would _like_ the `a.open("filename");` call to be an error, if I could figure out a way to make that happen. Logically it should be considered odd to open a file and then immediately destroy the `ofstream` object, at least by default. If it were actually desired, such as if it somehow generated a file system side effect (e.g., that just opening and immediately closing the file might trigger notifications to any programs hooked into file system notifications, which might include this program?), and in that case I think it would be beneficial for the code to have an explicit discard to highlight that we're doing something unusual (plus hopefully a comment about why, documenting the intended side effect outside the program for example).

Similarly, I came up with about the same `string` example Johel pointed out:

```
// minor variation on your example above
#include <string>
main: () = {
    a: std::string = ();
    a.append("filename");
}
```

I would prefer this also to be an error, because why append to a string and then not use that value?

The current Cpp2 status quo is that such cases are errors, but mostly only if the type being used was authored in Cpp2 (else if authored in Cpp1 it would have had to `&`-qualify the function or similar).

What do you think of that?

**Reply by tomFlowee** [reply]
> What do you think of that?

I think that is a really bad idea since it breaks a large chunk of C++ existing functionality. Your talk in 2023 cppcon wasn't exactly "developers developers developers", but it did repeat the same amount of times that backwards compatibility and drop-in-compatibility are the number one priority.

Yes, you may come up with a better way to do RAII (mutex guards, as a quick example), but the old design must keep working. That isn't really worth a debate.
Similarly to the examples I've given, calling a non-const method and then not doing anything with the object does not imply that said class is "dropped".
I'll repeat the usecase of a network-enabled class that gets a hostname / port and is from then on fire-and-forget.

The C++ scoping is shared with practically all popular languages out there, you're effectively trying to change that. And I don't like that either.
The example was given of a usage of 'foo' after a return of foo. This would be avoided, the claim was. But if you look at the actual code that isn't even true. It works just fine. (see example lower on this post, just below "current design has issues")

# Its the move constructor!

Using std::move has no effect on a POD object. Your destructor will keep on working just fine on the old one, and in actual fact your code that does a move on the last usage of a variable has no effect on that destructor still being called at the end of the scope. C++ is still C++.

But here is the important point, where this does make a difference is when the class has a move constructor. I maintain a library and there is a class in there which would misbehave on a copy constructor. So we made sure it **only** has a move constructor.

```
class A {
NetworkConnection m_con;
void foo() {
   auto con = network.createConnection(hostname, port);
   con.setCallback(blablabla);
   con.connect();
   m_con = std::move(con);
}
};
```

This is the general way of using said class, and the std::move there is mandatory. It helps people understand that the old one is no longer useful after that.

A cpp2 equivalent can skip writing the std::move, because it is implied by your compiler. Neat! :-)
The beautiful part is that if I actually do use the 'con' after that line, stuff stops compiling because, as I wrote, the std::move in my code example is mandatory. And your compiler would stop inserting it if I do something with `con` later.

So, this is to say that I think you've been looking at this whole thing wrong.

You're not protecting anyone by disallowing a non-const method to be called as the last thing on a POD object because there are millions of cases where this is the correct thing to do. It is not about the calling of non-const methods. It's not even about POD objects. Red Herring. Not going to get you what you want. In actual fact not giving you what you're after.

No, you want to have more classes that remove the copy constructor and use the move constructor: `Foo::Foo(Foo && o);`
because using a move constructor has the effect you are after.

```
class Foo {
public: Foo::Foo(Foo && o);
 Foo::Foo(Foo&) = delete;
 Foo::Foo();

void bar(); 
};

main: () = {
  foo: Foo = ();
  foo.bar();
  b: Foo = foo;
  // foo.bar(); 
}
```

This should work, but if you uncomment that last 'foo' then it will stop compiling.

Whereas your current design has issues:

```
class A {
  ~A() {
    std::cout << "~A";
  }
  void b() {}
};

main: () = {
  a: A = ();
  a.b();
  std::cout << "end of scope";
}
```

This shows that the destructor is still called after the `b()`, on the old instance, so you didn't really accomplish anything with that implicit `std::move`. The destructor print is done after the 'end of scope' one.

You can adjust this to use std::ofstream since that too is movable. And thus you'd get errors if you use the instance after it was assigned.

```
#include <fstream>
main: (args) -> int = {
    a: std::ofstream = ();
    a.open("filename");
    b: std::ofstream = a;
    b.close();
    // a.close();
}
```

Uncomment that last line and it stops compiling.
And THAT is useful.

**Reply by JohelEGP** [reply]
I already clarified that discarding is orthogonal to destruction: <https://github.com/hsutter/cppfront/discussions/901#discussioncomment-7912441>.

> Its the move constructor!

No, [we want more regular types](https://github.com/hsutter/cppfront/issues/815#issuecomment-1826432741).
Cpp2 double downs on Cpp1's value semantics.

> but it did repeat the same amount of times that backwards compatibility and drop-in-compatibility are the number one priority.

> but the old design must keep working

Writing in Cpp2 is generally better by default.
It's more safe, terse, and descriptive (e.g., with the parameter directions `in`, `inout`, etc.).

This highlighting of data flow is one of the things Cpp2 brings to the table.
Cpp1 also broke backwards compatibility with C with its stronger type system.

> calling a non-const method and then not doing anything with the object does not imply that said class is "dropped"

You're right, and in Cpp2, you confirm that with an explicit discard.
I understand that it can be seen as a net negative if you need it a lot in your code base.
That would certainly highlight the special behavior of such statements.

**Reply by tomFlowee** [reply]
> > Its the move constructor!

> No, we want more regular types.

Curious about why you start it with "no". There is no disagreement here... Notice I didn't make some suggestions, like auto-generate move semantics more than already done. So, it's the move semantics that are relevant here!

> This highlighting of data flow is one of the things Cpp2 brings to the table.

I think if you read the post you'll see that it doesn't actually work. Which is the reason why it keeps on coming back in bugreports.
Adding a std::move around a plain old data object **has no effect**. The generated code even honors the lifetime of objects till the end of the current scope.
The effects you guys say you want works and gives you benefit if the object disabled copy constructor and is movable. Hence, the above reply.

I showed the code examples, I highlighted the cases where it is beneficial as well.

> This highlighting of data flow is one of the things Cpp2 brings to the table.

And you can have this for stuff that needs it. Like a movable object. This is where actual bugs appear on misuse. And you catch them. Problem solved. Now lets remove the stray ampersand on the generated method signature. :man_shrugging:

**Reply by JohelEGP** [reply]
> > > Its the move constructor!
>
> > No, we want more regular types.
>
> Curious about why you start it with "no". There is no disagreement here... Notice I didn't make some suggestions, like auto-generate move semantics more than already done. So, it's the move semantics that are relevant here!

> I showed the code examples, I highlighted the cases where it is beneficial as well.
>
> > This highlighting of data flow is one of the things Cpp2 brings to the table.
>
> And you can have this for stuff that needs it. Like a movable object.

You want to not have to explicitly discard the "out" aspect of a last use of a variable.
You're suggesting that what we really want is to have more move-only types.
But that would mean that regular types,
for which discarding the "out" aspect on last use makes no sense,
would stop being diagnosed in Cpp2 code.

> > This highlighting of data flow is one of the things Cpp2 brings to the table.
>
> I think if you read the post you'll see that it doesn't actually work. Which is the reason why it keeps on coming back in bugreports.

I don't think bug reports equates not working.
Maybe it's just a combination of a teaching issue and adaptability.

> Adding a std::move around a plain old data object has no effect. The generated code even honors the lifetime of objects till the end of the current scope.

That's just an implementation detail.
The Cpp1 code doesn't need that `static_cast<void>(std::move(x));`.
Maybe it should just lower to a comment in Cpp2 code: `/* _ = x; */`.

**Reply by JohelEGP** [reply]
Or `CPP2_DISCARD(x);`.

**Reply by tomFlowee** [reply]
> You want to not have to explicitly discard the "out" aspect of a last use of a variable.

no, that is incorrect.

I"m claiming that calling a non-const method on an object does not create an 'out' **in that calling context**. It is being used, but not in the place where the variable is owned. Core example, in main().

I'm going to claim that discarding a variable makes scope VERY confusing. Is at best a workaround to a claim that something is an error which at best is a warning and many times is just normal correct usage.  
And I've been claiming various times that the real bugs are being caught already in movable objects misusage.

> You're suggesting that what we really want is to have more move-only types.

I literally said the opposite in the reply you quoted.

**Reply by JohelEGP** [reply]
Right, sorry.
What I said is the effect of what applying your model would be.

> > You're suggesting that what we really want is to have more move-only types.
> 
> I literally said the opposite in the reply you quoted.

Well, I'm confused.

**Reply by tomFlowee** [reply]
> What I said is the effect of what applying your model would be.

I didn't propose any model. I observed that the current setup is very good at catching mistakes for a certain type of classes. Movable ones. Of which there are plenty. So the idea behind this is fine.

What I also observed is that the wish to detect unused variables and error out on them has a lot of issues, to which I observed that the idea at the core doesn't make sense since the compiler can't possibly know if a certain thing is correct or not.
And after programming with cppfront for 2 days, I also observe that it is much worse than a naggy IDE when you can't write code incrementally with compiles in the meantime.

So, my conclusion is that the design works fine but there is the WEIRD adding of an ampersand to the function declarations which should be removed for code that is meant to be compatible with cpp.

**Comment by realgdman** [comment]
I'd say its correct, and API of `connect` need to be improved. To something like streams or files, where after operation you have to assign returned value and check if connection/subscription was successful.
```
status := a.connect("bar");
if(status.is_ok()) ... 
```
Then, discarding returned value is actually meaninful error.

**Comment by JohelEGP** [comment]
This is the code from <https://github.com/hsutter/cppfront/issues/900#issuecomment-1865033787>:
```Cpp2
QCoreApplication app;
Client client;
client.connect(hostname, post);
return app.exec();
```
The relation between `app` and `client` is somehow implicit, perhaps through global state.
That doesn't play well with the local analysis of lifetimes.

**Comment by SebastianTroy** [comment]
Could perhaps evaluate to a boolean so you can "if (not a.connect()) ;" unless you're thinking there will be extra optional error information in the type?


On 23 December 2023 18:04:52 realgdman ***@***.***> wrote:

I'd say its correct, and API of connect need to be improved. To something like streams or files, where after operation you have to assign returned value and check if connection/subscription was successful.

status := a.connect("bar");
if(status.is_ok()) ...

**Comment by SebastianTroy** [comment]
I'd argue that the lifetime there is very explicit, you've created an object on the stack and returned from the function, what else would/should a programmer expect? I personally don't want a language with lifetime extension via spooky action at a distance!


On 23 December 2023 18:21:52 Johel Ernesto Guerrero Peña ***@***.***> wrote:

This is the code from #900 (comment)<https://github.com/hsutter/cppfront/issues/900#issuecomment-1865033787>:

QCoreApplication app;
Client client;
client.connect(hostname, post);
return app.exec();


The relation between app and client is somehow implicit, perhaps through global state.
That doesn't play well with the local analysis of lifetimes.

**Comment by tomFlowee** [comment]
For people that don't have experience with Qt, it uses asynchronous setup for things like networking.
So a 'connect' will instantly return, while at the same time set into action the low level stuff to open a connection.

When the async connect succeeds (or hard fails) you get a callback. At some event in the future.
Event handler-callbacks are the entire reason for there being a "client" **object**.

As a whole this means that the 'connect' instantly returns and since the 'app.exec()' is running the event-queue any sort of callbacks are (quite deep in the stack) run still in main(). Specifically, the 'client' object is still being used for those callbacks and all the things a client would actually do after the connection actually established.

But obviously no compiler could possibly know about any of that via some simple static analisis. For starters, the entire system I just described is provided as a dynamic library. Without debug symbols.

The code I linked somewhere in this thread actually calls a server and fetches data to look through it. It runs for about an hour with about a million callbacks being processed on the event queue before all relevant data has been found and it exits using the async method of Qt, calling the static `QCoreApplication::quit()` method that makes that 'exec()' call return.
